File: app/src/main/java/com/vladimir_x/easyenglishlearn/category_select/CategoryViewModel.java
package com.vladimir_x.easyenglishlearn.category_select;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

import com.vladimir_x.easyenglishlearn.App;
import com.vladimir_x.easyenglishlearn.R;
import com.vladimir_x.easyenglishlearn.SingleLiveEvent;
import com.vladimir_x.easyenglishlearn.db.WordDao;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

import static com.vladimir_x.easyenglishlearn.Constants.EMPTY_STRING;

public class CategoryViewModel extends ViewModel {

    private WordDao mRepository;
    private LiveData<List<String>> mCategoriesLiveData;
    private SingleLiveEvent<String> mEditCategoryLiveData;
    private SingleLiveEvent<String> mRemoveDialogLiveData;
    private SingleLiveEvent<String> mOpenCategoryLiveData;
    private SingleLiveEvent<Void> mRemoveCategoryLiveData;
    private SingleLiveEvent<Integer> mMessageLiveData;
    private CompositeDisposable mDisposable;


    public CategoryViewModel() {
        mRepository = App.getInstance().getDatabase().wordDao();
        mCategoriesLiveData = mRepository.getAllCategories();
        mEditCategoryLiveData = new SingleLiveEvent<>();
        mRemoveDialogLiveData = new SingleLiveEvent<>();
        mOpenCategoryLiveData = new SingleLiveEvent<>();
        mRemoveCategoryLiveData = new SingleLiveEvent<>();
        mMessageLiveData = new SingleLiveEvent<>();
        mDisposable = new CompositeDisposable();
    }

    public void onFabClick() {
        mEditCategoryLiveData.setValue(EMPTY_STRING);
    }

    public void onRemoveIconClick(String categoryName) {
        mRemoveDialogLiveData.setValue(categoryName);
    }

    public void onEditIconClick(String categoryName) {
        mEditCategoryLiveData.setValue(categoryName);
    }

    public void onRvItemClick(String categoryName) {
        mOpenCategoryLiveData.setValue(categoryName);
    }

    public void removeCategory(String categoryName) {
        Disposable disposable = Completable
                .fromAction(() -> mRepository.removeCategory(categoryName))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::showMessage);
        mRemoveCategoryLiveData.call();
        mDisposable.add(disposable);
    }

    public void cancelRemoving() {
        mRemoveCategoryLiveData.call();
    }

    private void showMessage() {
        mMessageLiveData.setValue(R.string.category_removed);
    }

    @Override
    protected void onCleared() {
        mDisposable.dispose();
    }

    LiveData<String> getOpenCategoryLiveData() {
        return mOpenCategoryLiveData;
    }

    LiveData<String> getEditCategoryLiveData() {
        return mEditCategoryLiveData;
    }

    LiveData<String> getRemoveDialogLiveData() {
        return mRemoveDialogLiveData;
    }

    LiveData<Void> getRemoveCategoryLiveData() {
        return mRemoveCategoryLiveData;
    }

    LiveData<List<String>> getCategoriesLiveData() {
        return mCategoriesLiveData;
    }

    LiveData<Integer> getMessageLiveData() {
        return mMessageLiveData;
    }
}

File: app/src/main/java/com/vladimir_x/easyenglishlearn/category_select/CategoryViewModel.kt
package com.vladimir_x.easyenglishlearn.category_select

import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.vladimir_x.easyenglishlearn.App
import com.vladimir_x.easyenglishlearn.Constants
import com.vladimir_x.easyenglishlearn.R
import com.vladimir_x.easyenglishlearn.SingleLiveEvent
import com.vladimir_x.easyenglishlearn.db.WordDao
import kotlinx.coroutines.launch

class CategoryViewModel : ViewModel() {
    private val repository: WordDao? = App.instance?.database?.wordDao()
    private val _editCategoryLiveData: SingleLiveEvent<String> = SingleLiveEvent()
    private val _removeDialogLiveData: SingleLiveEvent<String> = SingleLiveEvent()
    private val _openCategoryLiveData: SingleLiveEvent<String> = SingleLiveEvent()
    private val _removeCategoryLiveData: SingleLiveEvent<Unit> = SingleLiveEvent()
    private val _messageLiveData: SingleLiveEvent<Int> = SingleLiveEvent()
    val categoriesLiveData = MediatorLiveData<List<String>>()

    val openCategoryLiveData: LiveData<String?>
        get() = _openCategoryLiveData
    val editCategoryLiveData: LiveData<String?>
        get() = _editCategoryLiveData
    val removeDialogLiveData: LiveData<String?>
        get() = _removeDialogLiveData
    val removeCategoryLiveData: LiveData<Unit?>
        get() = _removeCategoryLiveData
    val messageLiveData: LiveData<Int?>
        get() = _messageLiveData

    init {
        viewModelScope.launch {
            categoriesLiveData.addSource(repository!!.getAllCategories()) {
                categoriesLiveData.value = it
            }
        }
    }

    fun onFabClick() {
        _editCategoryLiveData.value = Constants.EMPTY_STRING
    }

    fun onRemoveClick(categoryName: String?) {
        _removeDialogLiveData.value = categoryName
    }

    fun onEditClick(categoryName: String?) {
        _editCategoryLiveData.value = categoryName
    }

    fun onItemClick(categoryName: String?) {
        _openCategoryLiveData.value = categoryName
    }

    fun removeCategory(categoryName: String?) {
        viewModelScope.launch {
            repository?.removeCategory(categoryName)
            showMessage()
            _removeCategoryLiveData.call()
        }
    }

    fun cancelRemoving() {
        _removeCategoryLiveData.call()
    }

    private fun showMessage() {
        _messageLiveData.value = R.string.category_removed
    }
}